tiingo/wsclient.py

```diff
@@ -0,0 +1,95 @@
+        #  
+        #   # see https://api.tiingo.com/documentation/websockets/iex > Response for more info
+        #   "data":[] # an array containing trade information and a timestamp
+        #   
+        # }
+
+        print(msg)
+
+    subscribe = {
+            'eventName':'subscribe',
+            'authorization':'API_KEY_GOES_HERE',
+            #see https://api.tiingo.com/documentation/websockets/iex > Request for more info
+            'eventData': { 
+                'thresholdLevel':5
+          }
+    }
+    # notice how the object isn't needed after using it
+    # any logic should be implemented in the callback function 
+    TiingoWebsocketClient(subscribe,endpoint="iex",on_msg_cb=cb_fn)
+    while True:pass
+    '''
+
+    def __init__(self,config={},endpoint=None,on_msg_cb=None):
+        
+        self._base_url = "wss://api.tiingo.com"
+        self.config=config
+        
+        try:
+            api_key = self.config['authorization']
+        except KeyError:
+            api_key = os.environ.get('TIINGO_API_KEY')
+            self.config.update({"authorization":api_key})
+
+        self._api_key = api_key
+        if not(api_key):
+            raise RuntimeError("Tiingo API Key not provided. Please provide"
+                               " via environment variable or config argument."
+                               "Notice that this config dict takes the API Key as authorization ")
+
+        self.endpoint = endpoint
+        if not (self.endpoint=="iex" or self.endpoint=="fx" or self.endpoint=="crypto"):
+            raise AttributeError("Endpoint must be defined as either (iex,fx,crypto) ")
+        
+        self.on_msg_cb = on_msg_cb
+        if not self.on_msg_cb:
+            raise MissingRequiredArgumentError("please define on_msg_cb It's a callback that gets called when new messages arrive "
+                                          "Example:"
+                                          "def cb_fn(msg):"
+                                          "    print(msg)")
+
+        websocket.enableTrace(False)
+        
+        ws = websocket.WebSocketApp("{0}/{1}".format(self._base_url,self.endpoint),
+                              on_message = self.get_on_msg_cb(),
+                              on_error = self.on_error,
+                              on_close = self.on_close,
+                              on_open = self.get_on_open(self.config))
+        ws.run_forever()
+    
+    def get_on_open(self,config):
+        # the methods passed to websocketClient have to be unbounded if we want WebSocketApp to pass everything correctly 
+        # see websocket-client/#471
+        def on_open(ws):
+            ws.send(json.dumps(config))
+        return on_open
+
+    def get_on_msg_cb(self):
+        def on_msg_cb_local(ws,msg):
+            self.on_msg_cb(msg)
+            return 
+        return on_msg_cb_local
+
+    def on_error(ws, error):
```

Comment:
Let's make the first argument to `on_error` and `on_close` `self` rather than `ws`, to address the warning provided by LGTM

- https://lgtm.com/projects/g/hydrosquall/tiingo-python/rev/pr-ba63650c8fdd00a84b804ee3327857a1f001cbf6

Reply:
that's just going to be confusing. since these functions don't get ```self``` when they are called. and will just receive a ```ws``` instance. i think a better approach would be to just use ```*args``` so as not to confuse future maintainers.BUT i doubt that would stop the LGTM linter from throwing a warning

Reply:
I see your point about not wanting people to think that "self" refers to the Tiingo client instead of the `ws` instance - what do you think about:

- Leaving the function as is, but leaving a comment to explain to future maintainers why the variable is called `ws` on this line. This won't affect LGTM but will help maintainers
- Removing this function from the `TiingoClient` class and defining it as a standalone function above the class - this will solve the LGTM issue.

Reply:
let's go with the first solution. i think it's best for this case

Reply:
If you see this in time, can we also add a comment to suppress the warning?

https://lgtm.com/help/lgtm/alert-suppression#python

If not no worries, I'll add this before releasing to PyPi on Friday - this is to avoid future contributors from getting spammed by a warning they can ignore.